common/src/chain/config/mod.rs

```diff
@@ -454,10 +456,27 @@ impl ChainConfig {
         self.empty_consensus_reward_maturity_distance
     }
 
-    // TODO: this should be part of net-upgrades. There should be no canonical definition of PoW for any chain config
     #[must_use]
-    pub const fn get_proof_of_work_config(&self) -> PoWChainConfig {
-        PoWChainConfig::new(self.chain_type)
+    pub fn get_proof_of_work_config(&self) -> PoWChainConfig {
```

Comment:
Two points:

1. Why does this fix apply to PoW, not PoS?
2. This function was made const because this is called directly repeatedly in ChainConfig, which makes it expensive to create.

Reply:
For PoS in `calculate_target_required()`, in both `Threshold` or `Ongoing` we have either an `initial_difficulty` or `PoSChainConfig` to get the target value. However for PoW in `calculate_work_required(), we only have `initial_difficulty` for `Threshold` and nothing to go by for `Ongoing` and so have to trace back through the block index... but for Genesis since there we had no information, before this PR we returned a `ConsensusPoWError::GenesisCannotHaveOngoingDifficulty` error.

So instead of returning the error for PoW at Genesis-Ongoing, I initially created a new `PoWChainConfig` to return its `limit`... yet that above unit test was failing. Only after digging in I found that `PoW::new()` that returns a `PoWChainConfig` wasn't actually looking at the passed in `chain_config` and instead was just using the default values whereas my failing test wants to override these values.

Hopefully that all makes sense.

On 2, yeah I guess we need to keep it const... I'll have to rethink this one.

Reply:
I see.

Either find a way to make it const, or change the way it's used to avoid constantly reconstructing the configuration. I hope PoS doesn't have the same problem for its configuration that it just keeps getting reconstructed from a non-const function.

It's a bad design that was accepted at the very beginning of the project just to get things moving. I don't like depending on things that are ambiguous to performance and may or may not be optimized.

Reply:
For PoS we're not reconstructing it each time, but we are deconstructuring `pos_status`. But since we have the config for each arm, we're doing minimal work to get the target limit. Not `const`, but also too bad.

Also I don't think it's that much of an issue because these functions we're trying to optimise for here are only when the previous block is Genesis... 99% of the time it won't be, and we'll need to do heavy lifting by looking at the block index anyway, so I think we're focusing on optimising the wrong thing here, unless I'm wrong?

If performance is an issue, we can think about doing something like pre-calculating once at each epoch, or when blocks come in over the wire etc, and then using those values etc.